### PR TITLE
performance upgrades - primarily creation of packed fields for regridding 

### PR DESCRIPTION
**Pull request by mvertens**
### Description of changes
performance upgrades - primarily creation of packed fields for regridding

### Specific notes
This PR does several things.
1. The main focus is to do all on the mapping on packed fields. For this to happen a new data type, 
```fortran
type, public :: packed_data_type
     integer, allocatable :: fldindex(:) ! size of number of packed fields
     character(len=CS)    :: mapnorm     ! normalization for packed field
     type(ESMF_Field)     :: field_src    ! packed sourced field
     type(ESMF_Field)     :: field_dst    ! packed destination field
     type(ESMF_Field)     :: field_fracsrc
     type(ESMF_Field)     :: field_fracdst
  end type packed_data_type
```
where fldindex:
- corresponds to the mapping of the packed field ungridded dimension to the index in the source field bundle for the target field
- has a size of the ungridded dimension of the packed field (which equals the number of fields that will be mapped for a given mapping type and source/destination pair) PLUS the number of fields with ungridded dimensions times the number of ungridded dimensions for each of those fields.
The new mapping occurs by copying all relevant fields from a source field bundle into a packed, mapping the packed field and then copying the relevant array data from the packed field back into the mapped field bundle. The cost of the copying is extremely small.

2. It turns out that querying for the rank of a field is costly and seemed to account for a lot of the cost of `med_merge_mod`. As a result, queries for the rank have in most cases been replaced for queries for the size of the ungridded dimension of a field, thereby significantly lowering the cost of the model run. Also -  `med_methods_mod.F90` has been reduced to remove unused methods as well as to replace the rank calls in most places.
3. Implemented the option to map atm->ocn and atm-ice uv fields with cart3d based on introducing a new mapping type mappatch_cart3d. This is turned on via the nuopc.config variable `mapuv_with_cart3d`  which by default is` .false.`

Contributors other than yourself, if any:

CMEPS Issues Fixed (include github issue #):

Are changes expected to change answers?
 - [X] bit for bit - all compses other than X 
 - [ ] different at roundoff level
 - [x] more substantial - only X compsets (changes to the glc fields)

Any User Interface Changes (namelist or namelist defaults changes)?
 - [ ] Yes
 - [x] No

Testing performed if application target is CESM:(either UFS-S2S or CESM testing is required):
- [ ] (required) CIME_DRIVER=nuopc scripts_regression_tests.py
   - machines:
   - details (e.g. failed tests):
- [x] (required) CESM testlist_drv.xml
   - machines and compilers: cheyenne, intel
   - details (e.g. failed tests): No failed tests, compared to **oct26** baseline and generated **nov02** baselines in cmeps_baselines
- [ ] (optional) CESM prealpha test
   - machines and compilers
   - details (e.g. failed tests):

Testing performed if application target is UFS-S2S:
- [ ] (required) UFS-S2S testing
   - description:
   - details (e.g. failed tests):

Hashes used for testing:
- [x] CESM:
  - repository to check out: https://github.com/ESCOMP/CESM.git
  - branch: nuopc_dev
  - hash: 2740d0c
- [ ] UFS-S2S, then umbrella repostiory to check out and associated hash:
  - repository to check out:
  - branch:
  - hash:

### CODE REVIEW

File: mediator/med_phases_prep_glc_mod.F90
Comment (jedwards4b) [inline mediator/med_phases_prep_glc_mod.F90:911]: Is this comment still valid?  Don't see the referenced 'norm'

File: mediator/med_phases_prep_ocn_mod.F90
Comment (jedwards4b) [inline mediator/med_phases_prep_ocn_mod.F90:84]: should this be removed?

File: mediator/med_phases_prep_ocn_mod.F90
Comment (jedwards4b) [inline mediator/med_phases_prep_ocn_mod.F90:102]: Should this be removed?

**Review by jedwards4b — APPROVED** [review]

File: mediator/med_io_mod.F90
Comment (uturuncoglu) [inline mediator/med_io_mod.F90:529]: This assumes that all the fields in the field list is in same mesh. Right?

File: mediator/med_io_mod.F90
Comment (uturuncoglu) [inline mediator/med_io_mod.F90:601]: Why this is here. If I remember correctly Sa_topo (or maybe I am wrong) was not matching when it is compared with the cprnc. Is this control for it?

File: mediator/med_map_mod.F90
Comment (uturuncoglu) [inline mediator/med_map_mod.F90:856]: It would be nice to check those are allocated or not. It is valid for all allocate statements.

File: mediator/med_map_mod.F90
Comment (uturuncoglu) [inline mediator/med_map_mod.F90:874]: same for the deallocate statements. if those are already allocated then the deallocate can be called.

File: mediator/med_phases_ocnalb_mod.F90
Comment (uturuncoglu) [inline mediator/med_phases_ocnalb_mod.F90:97]: since all pointers init to null(), it would be nice to init this one with null().

File: mediator/med_phases_prep_ice_mod.F90
Comment (uturuncoglu) [inline mediator/med_phases_prep_ice_mod.F90:58]: init with null()

**Review by uturuncoglu — COMMENTED** [review]
@mvertens Let me test with the HAFS application before merge. I'll let you know about it. As we discussed with Jim yesterday, we might also want to update the PR template and include also testing with HAFS application. What do you think?

File: mediator/med_phases_prep_ice_mod.F90
Comment (uturuncoglu) [reply mediator/med_phases_prep_ice_mod.F90:58]: There are also other cases that pointiest are not set to null() but I did not mark all of them.

**Comment by DeniseWorthen** [comment]
Is this branch up to date w/ master? I don't see the recent changes in med_phases_aofluxes_mod.F90 where the 'first call' was moved into the init.

File: mediator/med_phases_prep_glc_mod.F90
Comment (mvertens) [reply mediator/med_phases_prep_glc_mod.F90:911]: I've added a clarifying statement.

File: mediator/med_phases_prep_ocn_mod.F90
Comment (mvertens) [reply mediator/med_phases_prep_ocn_mod.F90:84]: Good catch. I've removed this.

File: mediator/med_phases_prep_ocn_mod.F90
Comment (mvertens) [reply mediator/med_phases_prep_ocn_mod.F90:102]: Removed as well. Thanks.

File: mediator/med_io_mod.F90
Comment (mvertens) [reply mediator/med_io_mod.F90:529]: Yes that is correct. All fields in a given fieldbundle are assumed to be on the same mesh. That is an underlying assumption in CMEPS.

File: mediator/med_map_mod.F90
Comment (mvertens) [reply mediator/med_map_mod.F90:856]: This allocate just occurs at initialization - so Its not clear to me why it would not be allocated. But I can add a check statement.

Comment (mvertens) [reply mediator/med_map_mod.F90:856]: But will this require rerunning all of the tests again?

File: mediator/med_map_mod.F90
Comment (mvertens) [reply mediator/med_map_mod.F90:874]: You need this deallocate for npacked since you are allocating and deallocating it as you go through the mapindices. I don't ever see a case where you would trigger a deallocate for an unallocated array. I'll double check.

File: mediator/med_phases_ocnalb_mod.F90
Comment (mvertens) [reply mediator/med_phases_ocnalb_mod.F90:97]: Good point. I'm going to go through and set all pointers to null for consistency.

File: mediator/med_map_mod.F90
Comment (uturuncoglu) [reply mediator/med_map_mod.F90:856]: it is fine, we could do it later. I am testing it with HAFS now.

**Comment by rsdunlapiv** [comment]
@DeniseWorthen would it be possible to have before and after timings if you test this?  The work from @mvertens should speed up mapping for UFS S2S as well.

**Comment by DeniseWorthen** [comment]
> @DeniseWorthen would it be possible to have before and after timings if you test this? The work from @mvertens should speed up mapping for UFS S2S as well.

Sure; I'll take make a comparison.

File: mediator/med_io_mod.F90
Comment (mvertens) [reply mediator/med_io_mod.F90:601]: I don't think so. But frankly, I do not remember why it was put in. Maybe in the next PR or so I can address this question along with the other ones you raised. Can you please put in an issue for this.

File: mediator/med_map_mod.F90
Comment (mvertens) [reply mediator/med_map_mod.F90:856]: Okay - that would be great. Maybe you could raise an issue for this as well so that we don't forget to do it. Thanks for pointing this out.

File: mediator/med_phases_ocnalb_mod.F90
Comment (mvertens) [reply mediator/med_phases_ocnalb_mod.F90]: I just pushed back all local pointers being set to null(). That is what Denise is testing with - and I will redo the tests tonight.

**Comment by DeniseWorthen** [comment]
I have run 2 cases using the ufs-weather coupled control case at c384, 1/4 deg ocean and ice, comparing the perf update with the previous master. 

- The coupler restart files at 5d are b4b between these two runs

- Wall clock time in the two cases:
For the perf run: The total amount of wall time                        = 5748.449491
For the pre-perf run: The total amount of wall time                        = 5931.439371

Using this crude estimate, our 35d run might be ~20min faster.

**Comment by uturuncoglu** [comment]
@mvertens Just let you know that the HAFS regression test (datm+hycom) passed.

**Comment by mvertens** [comment]
@uturuncoglu @DeniseWorthen - when I reran the test suite last night I found unexpected issues with restart runs for some compsets. I think I found the problem and I'm rerunning the tests this morning.

**Comment by rsdunlapiv** [comment]
@DeniseWorthen thanks for reporting on the timing.  Those are very encouraging numbers!  @mark-a-potts should be aware for this for his analysis of UFS S2S performance.

**Review by uturuncoglu — APPROVED** [review]

**Comment by uturuncoglu** [comment]
@mvertens let me know if you need help for testing.

**Comment by mvertens** [comment]
@uturuncoglu @DeniseWorthen - I have fixed all the problems and tested again on my end and all the tests pass now are are bfb with oct28. Could you please test again and make sure that nothing changes for you. Sorry to make you do this twice - but this is a really big PR so I want to make sure that everything works correctly.

**Comment by DeniseWorthen** [comment]
It ran in nems_frac but didn't complete in nems_orig. I have a run here: /glade/scratch/worthen/hour2_escomp

**Review by DeniseWorthen — APPROVED** [review]

**Comment by DeniseWorthen** [comment]
All tests in ufs are b4b w/ the previous master for this PR.

**Comment by mvertens** [comment]
Please do not merge yet. My latest tests show unexpected answer changes for I2000. I also need to resolve the conflict in med_map_mod.F90 - not sure why this is there.

**Comment by mvertens** [comment]
So it looks like the oct28 baseline is incorrect for the test ERS_Vnuopc_Ld5_P144.f19_f19_mg17.I2000Clm50Sp.cheyenne_intel.clm-default
The test from this PR is bfb with both the oct27 and the oct26 baselines in cmeps_baselines.
I will generate a new baseline and then I think the PR will be ready to merge.

**Pull request merged**
